include/particles/particle.tcc

```diff
@@ -852,10 +855,21 @@ void mpm::Particle<Tdim>::apply_particle_velocity_constraints(unsigned dir,
   this->velocity_(dir) = velocity;
 }
 
+//! Return particle scalar data
+template <unsigned Tdim>
+inline double mpm::Particle<Tdim>::scalar_data(
+    const std::string& property) const {
+  return (this->scalar_properties_.find(property) !=
+          this->scalar_properties_.end())
+             ? this->scalar_properties_.at(property)()
+             : std::numeric_limits<double>::quiet_NaN();
+}
+
 //! Return particle tensor data
 template <unsigned Tdim>
-Eigen::VectorXd mpm::Particle<Tdim>::tensor_data(const std::string& property) {
-  return this->properties_.at(property)();
+inline Eigen::VectorXd mpm::Particle<Tdim>::tensor_data(
+    const std::string& property) const {
+  return this->tensor_properties_.at(property)();
```

Comment:
Shouldn't we return `nans` for tensorial quantities too? This may need us to split vector and tensorial properties.

Reply:
Done. I still keep the template `Tsize` in `particle_tensor_data` function in `mesh`. Should I remove the template and change default to 6? I am leaving it there so that if one day there is someone wants to output a nonsymmetric tensor with more than 6 independent components, we can slightly modify it.